File: backend/src/models/proposal_schema.py
"""
proposal_schema.py — Strict Pydantic Models for Final Proposal Schema
=======================================================================
Defines the strict schema enforced by Instructor during AI Orchestration
to guarantee 100% JSON compliance and zero hallucination/data loss.
"""

from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field, ConfigDict


class ProposalHotel(BaseModel):
    """Hotel accommodation details extracted from proposal/PDF."""
    model_config = ConfigDict(populate_by_name=True)

    name: str = Field(..., description="Exact hotel name verbatim")
    category: Optional[str] = Field(None, description="Star rating or accommodation type")
    price_per_night: Optional[float] = Field(None, description="Exact price per night if mentioned in document")
    location: Optional[str] = Field(None, description="City or area of the hotel")
    meal_plan: Optional[str] = Field(None, description="Meal plan e.g. CP, MAP, AP")
    inclusions: List[str] = Field(default_factory=list, description="Specific room/hotel inclusions")
    image_url: Optional[str] = Field("", description="Image URL assigned or extracted")


class ProposalActivity(BaseModel):
    """Activity or sightseeing excursion details."""
    model_config = ConfigDict(populate_by_name=True)

    name: str = Field(..., description="Exact activity name verbatim")
    duration: Optional[str] = Field(None, description="Duration e.g. 2 hours, Half day")
    timing: Optional[str] = Field(None, description="Timing e.g. 10:00 AM")
    price: Optional[float] = Field(None, description="Exact activity cost if mentioned")
    location: Optional[str] = Field(None, description="Location of activity")
    description: Optional[str] = Field(None, description="Full description verbatim")
    image_url: Optional[str] = Field("", description="Image URL assigned or extracted")


class ProposalTransfer(BaseModel):
    """Transfer or transportation segment."""
    model_config = ConfigDict(populate_by_name=True)

    transfer_type: Optional[str] = Field(None, alias="type", description="e.g. Airport Transfer, Inter-city Sightseeing")
    vehicle: Optional[str] = Field(None, description="Exact vehicle e.g. Innova Crysta, Tempo Traveller")
    from_location: Optional[str] = Field(None, alias="from", description="Origin location/point")
    to: Optional[str] = Field(None, description="Destination point")
    timing: Optional[str] = Field(None, description="Pickup timing e.g. 06:00 AM")
    price: Optional[float] = Field(None, description="Exact transfer cost if mentioned")
    notes: Optional[str] = Field(None, description="Additional transfer instructions or notes")


class ProposalMeal(BaseModel):
    """Meal arrangement details."""
    model_config = ConfigDict(populate_by_name=True)

    meal_type: Optional[str] = Field(None, alias="type", description="Breakfast, Lunch, Dinner, or Snacks")
    venue: Optional[str] = Field(None, description="Restaurant or hotel dining room name")
    cuisine: Optional[str] = Field(None, description="Type of cuisine")
    price: Optional[float] = Field(None, description="Cost per person/total if listed")
    notes: Optional[str] = Field(None, description="Special dietary notes or timing")
    image_url: Optional[str] = Field("", description="Image URL")


class ProposalDay(BaseModel):
    """Structured day-by-day itinerary schedule."""
    model_config = ConfigDict(populate_by_name=True)

    day_number: int = Field(..., description="Day index starting at 1")
    title: str = Field(..., description="Exact day title/headline")
    description: Optional[str] = Field("", description="Full verbatim day description text")
    sub_destination: Optional[str] = Field(None, description="Primary city/area for this day")
    schedule: Optional[str] = Field(None, description="General timing or schedule overview")
    hotels: List[ProposalHotel] = Field(default_factory=list)
    activities: List[ProposalActivity] = Field(default_factory=list)
    transfers: List[ProposalTransfer] = Field(default_factory=list)
    meals: List[ProposalMeal] = Field(default_factory=list)


class ProposalExtraSections(BaseModel):
    """Supplementary sections at end of proposal documents."""
    model_config = ConfigDict(populate_by_name=True, extra="allow")

    what_to_pack: Optional[str] = Field(None, description="Packing recommendations or essentials verbatim")
    visa_guidelines: Optional[str] = Field(None, description="Visa instructions or requirements")
    important_notes: Optional[str] = Field(None, description="General terms, notes, and remarks")
    damages: Optional[str] = Field(None, description="Damage or liability policy")
    cancellation_policy: Optional[str] = Field(None, description="Exact cancellation policy text")
    dos_and_donts: Optional[str] = Field(None, description="Do's and Don'ts guidelines")
    terms_of_payment: Optional[str] = Field(None, alias="payment", description="Payment schedule, advance deposit requirements and banking terms")
    terms_and_conditions: Optional[str] = Field(None, description="Exact terms and conditions verbatim")
    amendment: Optional[str] = Field(None, description="Amendment policy and communication charges verbatim")
    refund: Optional[str] = Field(None, description="Refund policy verbatim")
    about_transport: Optional[str] = Field(None, description="Specific transport rules, sector vehicle changes, or local taxi notes verbatim")
    arrival_requirements: Optional[str] = Field(None, description="Particulars or documents required on arrival (e.g. ID proof, photographs, age certificates)")


class FinalProposalSchema(BaseModel):
    """
    Final strict proposal schema enforced across extraction and AI orchestration.
    Guaranteeing zero schema drift and strict JSON compliance.
    """
    model_config = ConfigDict(populate_by_name=True)

    destination: str = Field(..., description="Primary destination name")
    sub_destinations: List[str] = Field(default_factory=list, description="All sub-destinations mentioned")
    overview: Optional[str] = Field("", description="Full verbatim introduction/overview text from page 1")
    duration_days: int = Field(..., description="Total itinerary duration in days")
    currency: str = Field("INR", description="ISO 4217 currency code detected e.g. INR, USD, EUR")
    total_price: Optional[float] = Field(None, description="Total package cost if explicitly mentioned")
    price_per_person: Optional[float] = Field(None, description="Per person cost if listed")
    days: List[ProposalDay] = Field(default_factory=list)
    inclusions: List[str] = Field(default_factory=list, description="Exact list of package inclusions")
    exclusions: List[str] = Field(default_factory=list, description="Exact list of package exclusions")
    extra_sections: ProposalExtraSections = Field(default_factory=ProposalExtraSections)

File: frontend/src/schemas/proposalSchema.js
import { z } from 'zod';

export const ProposalHotelSchema = z.object({
  name: z.string(),
  category: z.string().nullable().optional(),
  price_per_night: z.number().nullable().optional(),
  location: z.string().nullable().optional(),
  meal_plan: z.string().nullable().optional(),
  inclusions: z.array(z.string()).default([]),
  image_url: z.string().nullable().optional().default(""),
}).passthrough();

export const ProposalActivitySchema = z.object({
  name: z.string(),
  duration: z.string().nullable().optional(),
  timing: z.string().nullable().optional(),
  price: z.number().nullable().optional(),
  location: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
  image_url: z.string().nullable().optional().default(""),
}).passthrough();

export const ProposalTransferSchema = z.object({
  transfer_type: z.string().nullable().optional(),
  vehicle: z.string().nullable().optional(),
  from_location: z.string().nullable().optional(),
  to: z.string().nullable().optional(),
  timing: z.string().nullable().optional(),
  price: z.number().nullable().optional(),
  notes: z.string().nullable().optional(),
}).passthrough();

export const ProposalMealSchema = z.object({
  meal_type: z.string().nullable().optional(),
  venue: z.string().nullable().optional(),
  cuisine: z.string().nullable().optional(),
  price: z.number().nullable().optional(),
  notes: z.string().nullable().optional(),
  image_url: z.string().nullable().optional().default(""),
}).passthrough();

export const ProposalDaySchema = z.object({
  day_number: z.number().optional(), // optional because frontend might just use day or rely on index
  title: z.string().optional(),
  description: z.string().nullable().optional().default(""),
  sub_destination: z.string().nullable().optional(),
  schedule: z.string().nullable().optional(),
  hotels: z.array(ProposalHotelSchema).default([]),
  activities: z.array(ProposalActivitySchema).default([]),
  transfers: z.array(ProposalTransferSchema).default([]),
  meals: z.array(ProposalMealSchema).default([]),
}).passthrough(); // allows other fields UI might add, like UI state, day, etc.

export const ProposalExtraSectionsSchema = z.object({
  what_to_pack: z.string().nullable().optional(),
  visa_guidelines: z.string().nullable().optional(),
  important_notes: z.string().nullable().optional(),
  damages: z.string().nullable().optional(),
  cancellation_policy: z.string().nullable().optional(),
  dos_and_donts: z.string().nullable().optional(),
  terms_of_payment: z.string().nullable().optional(),
  terms_and_conditions: z.string().nullable().optional(),
  amendment: z.string().nullable().optional(),
  refund: z.string().nullable().optional(),
  about_transport: z.string().nullable().optional(),
  arrival_requirements: z.string().nullable().optional(),
}).catchall(z.any()); // allow arbitrary extra sections

export const FinalProposalSchema = z.object({
  destination: z.string().optional(), // making optional to handle partial frontend saves
  sub_destinations: z.array(z.string()).default([]),
  overview: z.string().nullable().optional().default(""),
  duration_days: z.number().optional(),
  currency: z.string().default("INR"),
  total_price: z.number().nullable().optional(),
  price_per_person: z.number().nullable().optional(),
  days: z.array(ProposalDaySchema).default([]),
  hotels: z.array(ProposalHotelSchema).default([]),
  inclusions: z.array(z.string()).default([]),
  exclusions: z.array(z.string()).default([]),
  extra_sections: ProposalExtraSectionsSchema.default({}),
  model_used: z.string().nullable().optional(),
}).passthrough(); // pass through DB fields like id, client_id, etc.
